src/CircuitOptimizer.cpp

```diff
@@ -527,4 +527,106 @@ namespace qc {
 		// remove resulting identities from circuit
 		removeIdentities(qc);
 	}
+	void CircuitOptimizer::removeFinalMeasurements(QuantumComputation &qc)
+	{
+		//remove final measurements in the circuit (for use in mapping tool)
+		auto it = qc.ops.rbegin();
+		while (it != qc.ops.rend())
+		{
+			if ((*it)->isNonUnitaryOperation())
+			{
+				if ((*it)->getType() == Measure)
+				{
+					qc.ops.erase((++it).base());
+				}
+				else
+				{
+					//First non measurement
+					break;
+				}
+			}
+			else
+			{ //not a measurement
+				break;
+			}
+		}
+	}
+
+	void CircuitOptimizer::decomposeSWAP(QuantumComputation &qc, bool isDirectedArchitecture)
+	{
+		//decompose SWAPS in three cnot and optionally in four H
+		auto it = qc.ops.begin();
+		while (it != qc.ops.end())
+		{
+			if ((*it)->isStandardOperation())
+			{
+				if ((*it)->getType() == qc::SWAP)
+				{
+					std::vector<unsigned short> targets = (*it)->getTargets();
+					it = qc.ops.erase(it);
+					it = qc.ops.insert(it, std::make_unique<StandardOperation>((*it)->getNqubits(), Control(targets[0]), targets[1], qc::X));
+					if (isDirectedArchitecture)
+					{
+						it = qc.ops.insert(it, std::make_unique<StandardOperation>((*it)->getNqubits(), targets[0], qc::H));
+						it = qc.ops.insert(it, std::make_unique<StandardOperation>((*it)->getNqubits(), targets[1], qc::H));
+						it = qc.ops.insert(it, std::make_unique<StandardOperation>((*it)->getNqubits(), Control(targets[0]), targets[1], qc::X));
+						it = qc.ops.insert(it, std::make_unique<StandardOperation>((*it)->getNqubits(), targets[0], qc::H));
+						it = qc.ops.insert(it, std::make_unique<StandardOperation>((*it)->getNqubits(), targets[1], qc::H));
+					}
+					else
+					{
+						it = qc.ops.insert(it, std::make_unique<StandardOperation>((*it)->getNqubits(), Control(targets[1]), targets[0], qc::X));
+					}
+					it = qc.ops.insert(it, std::make_unique<StandardOperation>((*it)->getNqubits(), Control(targets[0]), targets[1], qc::X));
```

Comment:
```suggestion
					unsigned short nqubits = (*it)->getNqubits();
					it = qc.ops.erase(it);
					it = qc.ops.insert(it, std::make_unique<StandardOperation>(nqubits, Control(targets[0]), targets[1], qc::X));
					if (isDirectedArchitecture)
					{
						it = qc.ops.insert(it, std::make_unique<StandardOperation>(nqubits, targets[0], qc::H));
						it = qc.ops.insert(it, std::make_unique<StandardOperation>(nqubits, targets[1], qc::H));
						it = qc.ops.insert(it, std::make_unique<StandardOperation>(nqubits, Control(targets[0]), targets[1], qc::X));
						it = qc.ops.insert(it, std::make_unique<StandardOperation>(nqubits, targets[0], qc::H));
						it = qc.ops.insert(it, std::make_unique<StandardOperation>(nqubits, targets[1], qc::H));
					}
					else
					{
						it = qc.ops.insert(it, std::make_unique<StandardOperation>(nqubits, Control(targets[1]), targets[0], qc::X));
					}
					it = qc.ops.insert(it, std::make_unique<StandardOperation>(nqubits, Control(targets[0]), targets[1], qc::X));
```

`(*it)->getNqubits()` is undefined behavior if the call to `erase` deletes the last operation, since then `it==qc.ops.end()` and good luck with accessing anything from there.